File: packages/testing/src/runner/cli-wrapper.ts
/**
 * @afd/testing - CLI Wrapper
 *
 * Executes CLI commands (afd call <command>) and captures structured output.
 * Supports both subprocess execution and in-process execution modes.
 */

import { spawn } from "node:child_process";
import type { CommandResult } from "@afd/core";
import type { StepError } from "../types/report.js";
import { createStepError } from "../types/report.js";

// ============================================================================
// Configuration Types
// ============================================================================

export interface CliConfig {
  /** Path to the AFD CLI executable (default: "afd") */
  cliPath?: string;

  /** Working directory for CLI execution */
  cwd?: string;

  /** Environment variables to pass to CLI */
  env?: Record<string, string>;

  /** Default timeout in milliseconds (default: 30000) */
  timeout?: number;

  /** MCP server URL for --connect option */
  serverUrl?: string;

  /** Whether to output verbose logging */
  verbose?: boolean;
}

export interface ExecuteOptions {
  /** Override timeout for this command */
  timeout?: number;

  /** Additional environment variables */
  env?: Record<string, string>;
}

// ============================================================================
// Execution Result Types
// ============================================================================

export interface ExecuteSuccess {
  success: true;
  result: CommandResult<unknown>;
  durationMs: number;
  stdout: string;
  stderr: string;
}

export interface ExecuteError {
  success: false;
  error: StepError;
  durationMs: number;
  stdout: string;
  stderr: string;
  exitCode?: number;
}

export type ExecuteResult = ExecuteSuccess | ExecuteError;

// ============================================================================
// CLI Wrapper Class
// ============================================================================

/**
 * Wrapper for executing AFD CLI commands.
 *
 * @example
 * ```typescript
 * const cli = new CliWrapper({ serverUrl: "http://localhost:3000/mcp" });
 * const result = await cli.execute("todo.create", { title: "Buy groceries" });
 * if (result.success) {
 *   console.log(result.result.data);
 * }
 * ```
 */
export class CliWrapper {
  private config: Required<Omit<CliConfig, "serverUrl">> & { serverUrl?: string };

  constructor(config: CliConfig = {}) {
    this.config = {
      cliPath: config.cliPath ?? "afd",
      cwd: config.cwd ?? process.cwd(),
      env: config.env ?? {},
      timeout: config.timeout ?? 30000,
      verbose: config.verbose ?? false,
      serverUrl: config.serverUrl,
    };
  }

  /**
   * Execute a command via the AFD CLI.
   *
   * @param command - Command name (e.g., "todo.create")
   * @param input - Input parameters as an object
   * @param options - Execution options
   */
  async execute(
    command: string,
    input?: Record<string, unknown>,
    options?: ExecuteOptions
  ): Promise<ExecuteResult> {
    const timeout = options?.timeout ?? this.config.timeout;
    const startTime = Date.now();

    // Build CLI arguments
    const args = this.buildArgs(command, input);

    if (this.config.verbose) {
      console.log(`[CLI] ${this.config.cliPath} ${args.join(" ")}`);
    }

    try {
      const { stdout, stderr, exitCode } = await this.spawn(args, timeout, options?.env);
      const durationMs = Date.now() - startTime;

      if (this.config.verbose) {
        console.log(`[CLI] exit=${exitCode} duration=${durationMs}ms`);
        if (stderr) console.log(`[CLI] stderr: ${stderr}`);
      }

      // Try to parse JSON output
      const parseResult = this.parseOutput(stdout, stderr, exitCode);

      if (parseResult.success) {
        return {
          success: true,
          result: parseResult.result,
          durationMs,
          stdout,
          stderr,
        };
      } else {
        return {
          success: false,
          error: parseResult.error,
          durationMs,
          stdout,
          stderr,
          exitCode,
        };
      }
    } catch (err) {
      const durationMs = Date.now() - startTime;
      const message = err instanceof Error ? err.message : String(err);

      return {
        success: false,
        error: createStepError("unknown", `CLI execution failed: ${message}`, {
          cause: err instanceof Error ? err : undefined,
        }),
        durationMs,
        stdout: "",
        stderr: message,
      };
    }
  }

  /**
   * Build CLI arguments for a command.
   */
  private buildArgs(command: string, input?: Record<string, unknown>): string[] {
    const args: string[] = ["call", command];

    // Add server URL if configured
    if (this.config.serverUrl) {
      args.push("--connect", this.config.serverUrl);
    }

    // Add JSON format for structured output
    args.push("--json");

    // Add input as JSON argument if provided
    if (input && Object.keys(input).length > 0) {
      args.push("--input", JSON.stringify(input));
    }

    return args;
  }

  /**
   * Spawn CLI process and capture output.
   */
  private spawn(
    args: string[],
    timeout: number,
    extraEnv?: Record<string, string>
  ): Promise<{ stdout: string; stderr: string; exitCode: number }> {
    return new Promise((resolve, reject) => {
      const proc = spawn(this.config.cliPath, args, {
        cwd: this.config.cwd,
        env: {
          ...process.env,
          ...this.config.env,
          ...extraEnv,
        },
        shell: process.platform === "win32",
        timeout,
      });

      let stdout = "";
      let stderr = "";

      proc.stdout.on("data", (data: Buffer) => {
        stdout += data.toString();
      });

      proc.stderr.on("data", (data: Buffer) => {
        stderr += data.toString();
      });

      proc.on("close", (code) => {
        resolve({ stdout, stderr, exitCode: code ?? 0 });
      });

      proc.on("error", (err) => {
        reject(err);
      });

      // Handle timeout
      const timeoutId = setTimeout(() => {
        proc.kill("SIGTERM");
        reject(new Error(`Command timed out after ${timeout}ms`));
      }, timeout);

      proc.on("close", () => {
        clearTimeout(timeoutId);
      });
    });
  }

  /**
   * Parse CLI output into a CommandResult.
   */
  private parseOutput(
    stdout: string,
    stderr: string,
    exitCode: number
  ): { success: true; result: CommandResult<unknown> } | { success: false; error: StepError } {
    // Try to extract JSON from stdout
    const trimmed = stdout.trim();

    // Look for JSON object in output (may have non-JSON prefix/suffix)
    const jsonMatch = trimmed.match(/\{[\s\S]*\}/);
    if (!jsonMatch) {
      // If exit code is 0 but no JSON, treat as success with string data
      if (exitCode === 0) {
        return {
          success: true,
          result: {
            success: true,
            data: trimmed || undefined,
          },
        };
      }

      // Non-zero exit with no JSON is an error
      return {
        success: false,
        error: createStepError("parse_error", stderr || stdout || "Command failed with no output"),
      };
    }

    try {
      const parsed = JSON.parse(jsonMatch[0]) as unknown;

      // Check if it's a CommandResult-like object
      if (typeof parsed === "object" && parsed !== null && "success" in parsed) {
        return {
          success: true,
          result: parsed as CommandResult<unknown>,
        };
      }

      // Wrap raw data in a CommandResult
      return {
        success: true,
        result: {
          success: exitCode === 0,
          data: parsed,
        },
      };
    } catch (err) {
      return {
        success: false,
        error: createStepError(
          "parse_error",
          `Failed to parse JSON output: ${err instanceof Error ? err.message : String(err)}`
        ),
      };
    }
  }

  /**
   * Update configuration.
   */
  configure(config: Partial<CliConfig>): void {
    if (config.cliPath !== undefined) this.config.cliPath = config.cliPath;
    if (config.cwd !== undefined) this.config.cwd = config.cwd;
    if (config.env !== undefined) this.config.env = config.env;
    if (config.timeout !== undefined) this.config.timeout = config.timeout;
    if (config.serverUrl !== undefined) this.config.serverUrl = config.serverUrl;
    if (config.verbose !== undefined) this.config.verbose = config.verbose;
  }
}

// ============================================================================
// Factory Function
// ============================================================================

/**
 * Create a CLI wrapper with the given configuration.
 */
export function createCliWrapper(config?: CliConfig): CliWrapper {
  return new CliWrapper(config);
}

File: python/src/afd/testing/cli_wrapper.py
"""
CLI Wrapper for executing AFD commands via subprocess.

Port of packages/testing/src/runner/cli-wrapper.ts
"""

from __future__ import annotations

import asyncio
import json
import os
import re
import time
from dataclasses import dataclass, field
from typing import Any

from afd.testing.scenarios.types import StepError, create_step_error


@dataclass
class CliConfig:
	"""Configuration for the CLI wrapper."""

	cli_path: str = "afd"
	cwd: str | None = None
	env: dict[str, str] | None = None
	timeout: int = 30000
	server_url: str | None = None
	verbose: bool = False


@dataclass
class ExecuteOptions:
	"""Per-execution option overrides."""

	timeout: int | None = None
	env: dict[str, str] | None = None


@dataclass
class ExecuteSuccess:
	"""Successful CLI execution result."""

	success: bool  # Always True
	result: dict[str, Any]
	duration_ms: int
	stdout: str
	stderr: str


@dataclass
class ExecuteError:
	"""Failed CLI execution result."""

	success: bool  # Always False
	error: StepError
	duration_ms: int
	stdout: str
	stderr: str
	exit_code: int | None = None


ExecuteResult = ExecuteSuccess | ExecuteError


class CliWrapper:
	"""Wrapper for executing AFD CLI commands.

	Example:
		cli = CliWrapper(CliConfig(server_url="http://localhost:3000/mcp"))
		result = await cli.execute("todo-create", {"title": "Buy groceries"})
		if result.success:
			print(result.result["data"])
	"""

	def __init__(self, config: CliConfig | None = None) -> None:
		cfg = config or CliConfig()
		self._cli_path = cfg.cli_path
		self._cwd = cfg.cwd or os.getcwd()
		self._env: dict[str, str] = cfg.env or {}
		self._timeout = cfg.timeout
		self._server_url = cfg.server_url
		self._verbose = cfg.verbose

	async def execute(
		self,
		command: str,
		input: dict[str, Any] | None = None,
		options: ExecuteOptions | None = None,
	) -> ExecuteResult:
		"""Execute a command via the AFD CLI."""
		timeout = (options.timeout if options and options.timeout else None) or self._timeout
		start = time.monotonic()

		args = self._build_args(command, input)

		if self._verbose:
			print(f"[CLI] {self._cli_path} {' '.join(args)}")

		try:
			stdout, stderr, exit_code = await self._spawn(
				args, timeout, options.env if options else None
			)
			duration_ms = int((time.monotonic() - start) * 1000)

			if self._verbose:
				print(f"[CLI] exit={exit_code} duration={duration_ms}ms")
				if stderr:
					print(f"[CLI] stderr: {stderr}")

			parse_result = self._parse_output(stdout, stderr, exit_code)

			if parse_result[0]:
				return ExecuteSuccess(
					success=True,
					result=parse_result[1],
					duration_ms=duration_ms,
					stdout=stdout,
					stderr=stderr,
				)
			else:
				return ExecuteError(
					success=False,
					error=parse_result[1],
					duration_ms=duration_ms,
					stdout=stdout,
					stderr=stderr,
					exit_code=exit_code,
				)
		except Exception as err:
			duration_ms = int((time.monotonic() - start) * 1000)
			message = str(err)
			return ExecuteError(
				success=False,
				error=create_step_error("unknown", f"CLI execution failed: {message}"),
				duration_ms=duration_ms,
				stdout="",
				stderr=message,
			)

	def configure(self, **kwargs: Any) -> None:
		"""Update configuration."""
		if "cli_path" in kwargs:
			self._cli_path = kwargs["cli_path"]
		if "cwd" in kwargs:
			self._cwd = kwargs["cwd"]
		if "env" in kwargs:
			self._env = kwargs["env"]
		if "timeout" in kwargs:
			self._timeout = kwargs["timeout"]
		if "server_url" in kwargs:
			self._server_url = kwargs["server_url"]
		if "verbose" in kwargs:
			self._verbose = kwargs["verbose"]

	def _build_args(self, command: str, input: dict[str, Any] | None = None) -> list[str]:
		"""Build CLI arguments for a command."""
		args = ["call", command]

		if self._server_url:
			args.extend(["--connect", self._server_url])

		args.append("--json")

		if input and len(input) > 0:
			args.extend(["--input", json.dumps(input)])

		return args

	async def _spawn(
		self,
		args: list[str],
		timeout_ms: int,
		extra_env: dict[str, str] | None = None,
	) -> tuple[str, str, int]:
		"""Spawn CLI process and capture output."""
		env = {**os.environ, **self._env}
		if extra_env:
			env.update(extra_env)

		proc = await asyncio.create_subprocess_exec(
			self._cli_path,
			*args,
			cwd=self._cwd,
			env=env,
			stdout=asyncio.subprocess.PIPE,
			stderr=asyncio.subprocess.PIPE,
		)

		try:
			stdout_bytes, stderr_bytes = await asyncio.wait_for(
				proc.communicate(), timeout=timeout_ms / 1000
			)
		except asyncio.TimeoutError:
			proc.kill()
			await proc.wait()
			raise TimeoutError(f"Command timed out after {timeout_ms}ms") from None

		stdout = stdout_bytes.decode("utf-8", errors="replace") if stdout_bytes else ""
		stderr = stderr_bytes.decode("utf-8", errors="replace") if stderr_bytes else ""
		exit_code = proc.returncode or 0

		return stdout, stderr, exit_code

	def _parse_output(
		self, stdout: str, stderr: str, exit_code: int
	) -> tuple[bool, Any]:
		"""Parse CLI output into a CommandResult or StepError.

		Returns (True, result_dict) on success, (False, StepError) on failure.
		"""
		trimmed = stdout.strip()

		json_match = re.search(r"\{[\s\S]*\}", trimmed)
		if not json_match:
			if exit_code == 0:
				return (True, {"success": True, "data": trimmed or None})
			return (
				False,
				create_step_error("parse_error", stderr or stdout or "Command failed with no output"),
			)

		try:
			parsed = json.loads(json_match.group(0))

			if isinstance(parsed, dict) and "success" in parsed:
				return (True, parsed)

			return (True, {"success": exit_code == 0, "data": parsed})
		except json.JSONDecodeError as err:
			return (
				False,
				create_step_error("parse_error", f"Failed to parse JSON output: {err}"),
			)


def create_cli_wrapper(config: CliConfig | None = None) -> CliWrapper:
	"""Create a CLI wrapper with the given configuration."""
	return CliWrapper(config)
